### PR TITLE
Add a loading bar for tools

### PR DESCRIPTION
**Pull request by nsarrazin**
This uses the prometheus metrics, takes the 90th percentile to estimate the size of the loading bar with a default to 15s.

The bar finishes quickly if done early, or hangs close to the end if late.

https://github.com/huggingface/chat-ui/assets/25119303/6476bbce-cb07-47d6-be14-1bdd5743872c

### CODE REVIEW

**Comment by mishig25** [comment]
image

in the img above, once the img was produced, shouldn't the bar be finished as well?

**Comment by nsarrazin** [comment]
Uh I think the video I attached is loading poorly on the web, will let you try locally if you can to see if the behaviour works ok for you there

**Comment by Saghen** [comment]
Looks great but maybe we shouldn't animate to completion once the tool returns. Also, the estimated time can be known ahead of time for some tools (web search is 5-10s, url fetch is 2-4s, gradio spaces have an existing estimate). It might be nice to let the tools define their own estimates dynamically

File: src/lib/components/chat/ToolUpdate.svelte [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+			loadingBarEl.classList.remove("hidden");
+			isShowingLoadingBar = true;
+			loadingBarEl.animate([{ width: "0%" }, { width: "100%" }], {
+				duration: availableTools.find((tool) => tool.name === toolName)?.timeToUseMS,
+				fill: "forwards",
+			});
+		}
+	});
+
+	// go to 100% quickly if loading is done
```
Comment (gary149) [inline src/lib/components/chat/ToolUpdate.svelte:38]: not sure this works

**Review by gary149 — APPROVED** [review]

**Comment by mishig25** [comment]
give me like 15mins, doing one more check/review

File: src/routes/+layout.server.ts [hunk 2]
```diff
@@ -172,6 +174,9 @@ export const load: LayoutServerLoad = async ({ locals, depends }) => {
 				description: tool.description,
 				isOnByDefault: tool.isOnByDefault,
 				isLocked: tool.isLocked,
+				timeToUseMS:
+					toolUseDuration.find((el) => el.labels.tool === tool.name && el.labels.quantile === 0.9)
+						?.value ?? 15000,
```
Comment (mishig25) [inline src/routes/+layout.server.ts:179]:
```suggestion
						?.value ?? 15_000,
```

**Review by mishig25 — APPROVED** [review]
LGTM! Made a small suggestion https://github.com/huggingface/chat-ui/pull/1191 (up to you to accept or not 👍 )

**Comment by nsarrazin** [comment]
Accepted the suggestion, also made sure the loading bar goes til the end properly. Will add estimates from gradio API in a follow up PR if they are returned

File: src/lib/components/chat/ToolUpdate.svelte [hunk 3]
```diff
@@ -64,7 +70,7 @@
 		>
 			<div
 				bind:this={loadingBarEl}
-				class="absolute -m-1 hidden h-full w-full rounded-l-lg bg-purple-500/5 transition-all dark:bg-purple-500/10"
+				class="absolute -m-1 hidden h-full w-[calc(100%+1rem)] rounded-lg bg-purple-500/5 transition-all dark:bg-purple-500/10"
```
Comment (gary149) [inline src/lib/components/chat/ToolUpdate.svelte:73]: looked more like a progress bar without the rounded lg on the right imo (you could add an overflow-hidden on the parent instead).

Comment (nsarrazin) [reply src/lib/components/chat/ToolUpdate.svelte:73]: agreed, working on bringing it back but fighting with the padding :laughing:

Comment (gary149) [reply src/lib/components/chat/ToolUpdate.svelte:73]: it was fine as it was imo :D even if it wasn't taking the full width

**Pull request merged**